lib/publiccloud/openstack_client.pm

```diff
@@ -0,0 +1,78 @@
+# SUSE's openQA tests
+#
+# Copyright 2021 SUSE LLC
+# SPDX-License-Identifier: FSFAP
+
+# Summary: Helper class for OpenStack connection and authentication
+#
+# Maintainer: qa-c team <qa-c@suse.de>
+
+package publiccloud::openstack_client;
+use Mojo::Base -base;
+use testapi;
+use utils qw(file_content_replace);
+
+use constant OS_CONFIG_FILE => '/root/.config/openstack/clouds.yaml';
+use constant CLOUD_INIT_FILE => '/root/mn_jeos.cloud-init';
+
+has auth_uri => sub { get_required_var('PUBLIC_CLOUD_OPENSTACK_AUTH_URI') };
+has project_name => sub { get_required_var('PUBLIC_CLOUD_OPENSTACK_PROJECT_NAME') };
+has project_id => sub { get_required_var('PUBLIC_CLOUD_OPENSTACK_PROJECT_ID') };
+has region => sub { get_var('PUBLIC_CLOUD_REGION', 'CustomRegion') };
+has username => sub { get_var('PUBLIC_CLOUD_USER', 'sles') };
+
+sub _check_credentials {
+    my ($self) = @_;
+    my $max_tries = 6;
+
+    for my $i (1 .. $max_tries) {
+        my $out = script_output('openstack image list', 300, proceed_on_failure => 1);
+        record_info('cli check', $out);
+        # There are multiple possible error messages, e.g.:
+        #   The request you have made requires authentication
+        #   Failed to discover available identity versions
+        #   Missing value auth-url required for auth plugin password
+        #   Cloud mycloud was not found
+        #   Could not find a suitable TLS CA certificate bundle, invalid path
+        # But if the command succeeds, it will return an ASCII table with 3 columns: ID, Name, Status
+        return 1 if ($out =~ /\| ID.*| Name.*| Status.*|/m);
+        sleep 30;
+    }
+    return;
+}
+
+sub init {
+    my ($self, %params) = @_;
+
+    assert_script_run('mkdir -p /root/.config/openstack');
+    assert_script_run('curl ' . data_url("jeos/clouds.yaml") . ' -o ' . OS_CONFIG_FILE);
+    assert_script_run('curl ' . data_url("jeos/mn_jeos.cloud-init") . ' -o ' . CLOUD_INIT_FILE);
+
+    # terraform complains if the certificate is in /usr/share/pki/trust/anchors/
+    #    "Error: Error parsing CA Cert from /usr/share/pki/trust/anchors/SUSE_Trust_Root.crt"
+    # but it works if it's in root directory
+    assert_script_run('cp /usr/share/pki/trust/anchors/SUSE_Trust_Root.crt.pem /root/SUSE_Trust_Root.crt.pem');
```

Comment:
I would create a symbolic links instead, so it remains valid after updates are applied (if ever)

Reply:
ok :)

Reply:
I tried with ln -s and it doesn't help...
```
# timeout 1800 terraform plan -no-color -var 'secgroup=icmp_ssh' -var 'keypair=openqa1644941182_0' -var 'image_id=f1f61591-d2fe-4263-81ef-117fbbe64655' -var 'instance_count=1' -var 'type=m1.small' -var 'region=CustomRegion' -var 'name=jlausuch-vm' -var 'tags={"openqa_ttl":8640300,"openqa_var_JOB_ID":5588,"openqa_var_NAME":"00005588-sle-15-SP3-JeOS-for-OpenStack-Cloud-x86_64-Build4.7.10-jeos-check_boot@64bit-virtio-vga"}' -out myplan; echo nA6CZ-$?-

Error: Error parsing CA Cert from /root/SUSE_Trust_Root.crt.pem

  on plan.tf line 18, in provider "openstack":
  18: provider "openstack" {

```
http://fromm.arch.suse.de/tests/5588